core/lib/zksync_core/src/vm_runner/storage.rs

```diff
@@ -0,0 +1,456 @@
+/// caught up). Always initialized as a `Postgres` variant and is then mutated into `Rocksdb`
+/// once RocksDB cache is caught up.
+/// 2. Loads data needed to re-execute the next unprocessed L1 batch.
+#[derive(Debug)]
+pub struct VmRunnerStorage<L: VmRunnerStorageLoader> {
+    pool: ConnectionPool<Core>,
+    l1_batch_params_provider: L1BatchParamsProvider,
+    chain_id: L2ChainId,
+    max_batches_to_load: u32,
+    state: Arc<RwLock<State>>,
+    rocksdb_cell: Arc<OnceCell<RocksDB<StateKeeperColumnFamily>>>,
+    _marker: PhantomData<L>,
+}
+
+#[derive(Debug)]
+struct State {
+    l1_batch_number: L1BatchNumber,
+    storage: BTreeMap<L1BatchNumber, StorageData>,
+}
+
+impl<L: VmRunnerStorageLoader> VmRunnerStorage<L> {
+    /// Creates a new VM runner storage using provided Postgres pool and RocksDB path.
+    pub async fn new(
+        pool: ConnectionPool<Core>,
+        rocksdb_path: String,
+        loader: L,
+        enum_index_migration_chunk_size: usize, // TODO: Remove
+        chain_id: L2ChainId,
+        max_batches_to_load: u32,
+    ) -> anyhow::Result<(Self, StorageSyncTask<L>)> {
+        let mut conn = pool.connection_tagged(L::name()).await?;
+        let l1_batch_params_provider = L1BatchParamsProvider::new(&mut conn)
+            .await
+            .context("Failed initializing L1 batch params provider")?;
+        drop(conn);
+        let rocksdb_cell = Arc::new(OnceCell::new());
+        let state = Arc::new(RwLock::new(State {
+            l1_batch_number: L1BatchNumber(0),
+            storage: BTreeMap::new(),
+        }));
+        let task = StorageSyncTask::new(
+            pool.clone(),
+            max_batches_to_load,
+            chain_id,
+            rocksdb_path,
+            enum_index_migration_chunk_size,
+            loader.clone(),
+            rocksdb_cell.clone(),
+            state.clone(),
+        )
+        .await?;
+        Ok((
+            Self {
+                pool,
+                l1_batch_params_provider,
+                chain_id,
+                max_batches_to_load,
+                state,
+                rocksdb_cell,
+                _marker: PhantomData,
+            },
+            task,
+        ))
+    }
+
+    async fn access_storage_inner(
+        &self,
+        stop_receiver: &watch::Receiver<bool>,
+        l1_batch_number: L1BatchNumber,
+    ) -> anyhow::Result<Option<PgOrRocksdbStorage<'_>>> {
+        if let Some(rocksdb) = self.rocksdb_cell.get() {
+            let state = self.state.read().await;
+            let mut conn = self
+                .pool
+                .connection_tagged(L::name())
+                .await
+                .context("Failed getting a Postgres connection")?;
+            let max_l1_batch = (state.l1_batch_number + self.max_batches_to_load - 1).min(
+                conn.blocks_dal()
+                    .get_sealed_l1_batch_number()
+                    .await?
+                    .unwrap_or_default(),
+            );
```

Comment:
I think max_l1_batch should be the last batch in `state`

Reply:
So, I forgot to add the `loop` sequence that sleeps and blocks until the data is available here which is fixed now. I can do it your way too, but then this will mean that `access_storage` is going to become fallible because the data is not available yet. Which is going to shift the retry loop to the user which I think would be worse.

Reply:
Nevermind the above, I have decided to get rid of blocking logic entirely and now `max_l1_batch` is indeed just the last batch in `state`